File: src/movieclaw_api/api/routes/downloaders.py
from __future__ import annotations

from fastapi import APIRouter, BackgroundTasks, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from movieclaw_api.schemas.downloader import (
    DownloaderPayload,
    DownloaderStatusUpdate,
    DownloaderView,
    DownloadSubmitPayload,
    DownloadSubmitView,
)
from movieclaw_api.schemas.response import ApiResponse, ok
from movieclaw_api.services.downloader_config import (
    DownloaderConfigService,
    verify_downloader,
)
from movieclaw_api.services.library_config import LibraryConfigService, derive_save_path
from movieclaw_api.services.torrent_submit import submit_torrent
from movieclaw_db.engine import get_session

router = APIRouter(prefix="/downloaders", tags=["downloaders"])


@router.post(
    "/submit",
    response_model=ApiResponse[DownloadSubmitView],
    summary="把一条搜索结果种子提交到默认下载器",
)
async def submit_download(
    payload: DownloadSubmitPayload,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[DownloadSubmitView]:
    """手动下载：带站点登录态取回 .torrent → 提交给默认下载器。

    保存目录三级取值：选了媒体库 → 库推导路径（主根/标题 (年份)，无标题时
    落库主根）；未选库 → 默认下载器配置的默认目录 → 下载器自身默认。
    提交幂等：种子已在下载器中不视为错误，data.already_exists=true。
    """
    library = None
    derived_path = None
    if payload.library_id is not None:
        library = await LibraryConfigService(session).get(payload.library_id)
        if payload.title:
            derived_path = derive_save_path(library, title=payload.title, year=payload.year)
        else:
            derived_path = library.primary_root

    result, row = await submit_torrent(
        session,
        site_id=payload.site_id,
        download_url=payload.download_url,
        tags=["movieclaw-manual"],
        save_path=derived_path,
    )
    assert row.id is not None  # 落库记录必有主键
    view = DownloadSubmitView(
        info_hash=result.info_hash,
        name=result.name,
        already_exists=result.already_exists,
        downloader_id=row.id,
        downloader_name=row.name,
        save_path=derived_path or row.save_path,
    )
    if result.already_exists:
        message = "该种子已在下载器中，未重复添加"
    elif library is not None:
        message = f"已提交到「{row.name}」，入库到「{library.name}」"
    else:
        message = f"已提交到「{row.name}」"
    return ok(view, message=message)


@router.get(
    "",
    response_model=ApiResponse[list[DownloaderView]],
    summary="列出已配置的下载器及连接状态",
)
async def list_downloaders(
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[list[DownloaderView]]:
    service = DownloaderConfigService(session)
    rows = await service.list_all()
    return ok([DownloaderView.from_model(r) for r in rows])


@router.get(
    "/{downloader_id}",
    response_model=ApiResponse[DownloaderView],
    summary="获取单个下载器详情",
)
async def get_downloader(
    downloader_id: int,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[DownloaderView]:
    service = DownloaderConfigService(session)
    return ok(DownloaderView.from_model(await service.get(downloader_id)))


@router.post(
    "",
    response_model=ApiResponse[DownloaderView],
    summary="添加一个下载器（保存后异步测试连接）",
)
async def create_downloader_config(
    payload: DownloaderPayload,
    background_tasks: BackgroundTasks,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[DownloaderView]:
    """保存下载器连接信息（状态置 pending），并在后台异步测试连通性。

    接口立即返回，前端可轮询 GET /downloaders/{id} 观察 status 变化：
    pending → verifying → active（可用）/ failed（见 last_error）。
    """
    service = DownloaderConfigService(session)
    row = await service.create(
        name=payload.name,
        client_type=payload.client_type,
        url=payload.url,
        username=payload.username,
        password=payload.password,
        save_path=payload.save_path,
        enabled=payload.enabled,
    )
    assert row.id is not None  # 落库后必有主键
    row = await service.start_verification(row.id)
    background_tasks.add_task(verify_downloader, row.id)
    return ok(DownloaderView.from_model(row), message="已保存，正在测试连接")


@router.put(
    "/{downloader_id}",
    response_model=ApiResponse[DownloaderView],
    summary="更新下载器配置（更新后重新测试连接）",
)
async def update_downloader_config(
    downloader_id: int,
    payload: DownloaderPayload,
    background_tasks: BackgroundTasks,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[DownloaderView]:
    service = DownloaderConfigService(session)
    await service.update(
        downloader_id,
        name=payload.name,
        client_type=payload.client_type,
        url=payload.url,
        username=payload.username,
        password=payload.password,
        save_path=payload.save_path,
        enabled=payload.enabled,
    )
    row = await service.start_verification(downloader_id)
    background_tasks.add_task(verify_downloader, downloader_id)
    return ok(DownloaderView.from_model(row), message="已更新，正在测试连接")


@router.patch(
    "/{downloader_id}/status",
    response_model=ApiResponse[DownloaderView],
    summary="启用 / 停用下载器",
)
async def set_downloader_status(
    downloader_id: int,
    payload: DownloaderStatusUpdate,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[DownloaderView]:
    service = DownloaderConfigService(session)
    row = await service.set_enabled(downloader_id, payload.enabled)
    return ok(DownloaderView.from_model(row))


@router.post(
    "/{downloader_id}/default",
    response_model=ApiResponse[DownloaderView],
    summary="设为默认下载器",
)
async def set_default_downloader(
    downloader_id: int,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[DownloaderView]:
    """把该下载器设为默认（一键下载不选目标时投给它），同时取消其他台的默认。

    注意：其他记录的 is_default 也会随之变化，前端应整体刷新列表。"""
    service = DownloaderConfigService(session)
    row = await service.set_default(downloader_id)
    return ok(DownloaderView.from_model(row), message="已设为默认下载器")


@router.post(
    "/{downloader_id}/verify",
    response_model=ApiResponse[DownloaderView],
    summary="手动重新测试连接",
)
async def reverify_downloader(
    downloader_id: int,
    background_tasks: BackgroundTasks,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[DownloaderView]:
    """手动触发一次连接测试（如下载器重启后想确认恢复）。

    行为：不存在 → 404；已在测试中 → 409；否则同步占位为 VERIFYING
    并在后台重新测试。"""
    service = DownloaderConfigService(session)
    row = await service.start_verification(downloader_id)
    background_tasks.add_task(verify_downloader, downloader_id)
    return ok(DownloaderView.from_model(row), message="已重新发起连接测试")


@router.delete(
    "/{downloader_id}",
    response_model=ApiResponse[dict],
    summary="删除下载器配置",
)
async def delete_downloader_config(
    downloader_id: int,
    session: AsyncSession = Depends(get_session),
) -> ApiResponse[dict]:
    service = DownloaderConfigService(session)
    await service.delete(downloader_id)
    return ok({}, message="已删除")

File: apps/web/lib/api/downloaders.ts
import { request } from "@/lib/http";

/** 后端统一响应信封（见 movieclaw_api.schemas.response.ApiResponse） */
interface ApiEnvelope<T> {
  success: boolean;
  code: string;
  message: string;
  data: T;
}

async function unwrap<T>(promise: Promise<ApiEnvelope<T>>): Promise<T> {
  return (await promise).data;
}

/** 已适配的下载器类型（与后端 movieclaw_db ClientType 对应）。 */
export type DownloaderClientType = "qbittorrent" | "transmission";

/** 连接验证状态（与站点配置共用同一状态机语义）。 */
export type DownloaderStatus = "pending" | "verifying" | "active" | "failed";

/** 已配置下载器的对外视图（见 schemas.downloader.DownloaderView，脱敏无密码）。 */
export interface ConfiguredDownloader {
  id: number;
  name: string;
  client_type: DownloaderClientType;
  url: string;
  username: string | null;
  save_path: string | null;
  enabled: boolean;
  /** 是否为默认下载器（一键下载不选目标时投给它） */
  is_default: boolean;
  status: DownloaderStatus;
  /** 是否可用 = 已启用且连接测试通过 */
  usable: boolean;
  /** 最近一次连接成功获取的版本号，如 "v5.0.2" */
  version: string | null;
  last_error: string | null;
  last_checked_at: string | null;
  created_at: string;
  updated_at: string;
}

/** 新增/更新下载器的请求体（见 schemas.downloader.DownloaderPayload）。 */
export interface DownloaderPayload {
  name: string;
  client_type: DownloaderClientType;
  url: string;
  username?: string | null;
  password?: string | null;
  save_path?: string | null;
  enabled?: boolean;
}

/** 列出所有已配置的下载器及连接状态。 */
export function listDownloaders(init?: RequestInit): Promise<ConfiguredDownloader[]> {
  return unwrap(request<ApiEnvelope<ConfiguredDownloader[]>>("/downloaders", init));
}

/** 获取单个下载器详情（用于轮询连接测试进度）。 */
export function getDownloader(id: number, init?: RequestInit): Promise<ConfiguredDownloader> {
  return unwrap(request<ApiEnvelope<ConfiguredDownloader>>(`/downloaders/${id}`, init));
}

/** 添加一个下载器（保存后后端异步测试连接）。 */
export function createDownloader(payload: DownloaderPayload): Promise<ConfiguredDownloader> {
  return unwrap(
    request<ApiEnvelope<ConfiguredDownloader>>("/downloaders", {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  );
}

/** 更新下载器配置（更新后后端重新测试连接）。 */
export function updateDownloader(
  id: number,
  payload: DownloaderPayload,
): Promise<ConfiguredDownloader> {
  return unwrap(
    request<ApiEnvelope<ConfiguredDownloader>>(`/downloaders/${id}`, {
      method: "PUT",
      body: JSON.stringify(payload),
    }),
  );
}

/** 启用 / 停用下载器。 */
export function setDownloaderEnabled(
  id: number,
  enabled: boolean,
): Promise<ConfiguredDownloader> {
  return unwrap(
    request<ApiEnvelope<ConfiguredDownloader>>(`/downloaders/${id}/status`, {
      method: "PATCH",
      body: JSON.stringify({ enabled }),
    }),
  );
}

/** 设为默认下载器。注意：其他条目的 is_default 会随之变化，调用后应整体刷新列表。 */
export function setDefaultDownloader(id: number): Promise<ConfiguredDownloader> {
  return unwrap(
    request<ApiEnvelope<ConfiguredDownloader>>(`/downloaders/${id}/default`, { method: "POST" }),
  );
}

/** 手动重新测试一次连接。 */
export function reverifyDownloader(id: number): Promise<ConfiguredDownloader> {
  return unwrap(
    request<ApiEnvelope<ConfiguredDownloader>>(`/downloaders/${id}/verify`, { method: "POST" }),
  );
}

/** 删除下载器配置。 */
export function deleteDownloader(id: number): Promise<Record<string, never>> {
  return unwrap(
    request<ApiEnvelope<Record<string, never>>>(`/downloaders/${id}`, { method: "DELETE" }),
  );
}

/** 手动提交下载的请求体（见 schemas.downloader.DownloadSubmitPayload）。 */
export interface DownloadSubmitPayload {
  /** 种子所属站点 ID（TorrentHit.site_id） */
  site_id: string;
  /** 种子下载入口（TorrentHit.download_url） */
  download_url: string;
  /** 入库目标库（可选）：带上后保存目录由库推导（主根/标题 (年份)） */
  library_id?: number | null;
  /** 条目标题（推导条目子目录用；身份未确认时不要带） */
  title?: string | null;
  year?: number | null;
  /** 种子副标题（识别线索：中文片名/「全N集」帮扫描器收敛拼音命名种子） */
  subtitle?: string | null;
}

/** 手动提交下载的结果（见 schemas.downloader.DownloadSubmitView）。 */
export interface DownloadSubmitResult {
  info_hash: string | null;
  name: string;
  /** 种子提交前已存在于下载器（幂等，未重复添加） */
  already_exists: boolean;
  downloader_id: number;
  downloader_name: string;
  /** 实际使用的保存目录（null = 下载器自身默认目录） */
  save_path: string | null;
}

/**
 * 把一条搜索结果种子提交到默认下载器：后端带站点登录态取回 .torrent 再递交，
 * 保存目录用默认下载器配置的默认目录。失败抛 HttpError，message 为可读中文。
 */
export function submitTorrentDownload(
  payload: DownloadSubmitPayload,
): Promise<DownloadSubmitResult> {
  return unwrap(
    request<ApiEnvelope<DownloadSubmitResult>>("/downloaders/submit", {
      method: "POST",
      body: JSON.stringify(payload),
    }),
  );
}
